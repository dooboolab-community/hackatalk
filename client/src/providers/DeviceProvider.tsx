import * as Device from 'expo-device';

import React, {FC, useEffect, useState} from 'react';

import createCtx from '../utils/createCtx';

interface Context {
  deviceType: Device.DeviceType;
  setDeviceType: React.Dispatch<React.SetStateAction<Device.DeviceType>>;
}

const [useCtx, Provider] = createCtx<Context>();

const DeviceProvider: FC = ({children}) => {
  const [deviceType, setDeviceType] = useState<Device.DeviceType>(
    Device.DeviceType.PHONE,
  );

  useEffect(() => {
    let isMounted = true;

    Device.getDeviceTypeAsync().then((result) => {
      // Only change state if this component is still mounted.
      if (isMounted) setDeviceType(result);
    });

    return () => {
      isMounted = false;
    };
  }, []);

  return (
    <Provider
      value={{
        deviceType,
        setDeviceType,
      }}>
      {children}
    </Provider>
  );
};

const DeviceContext = {
  useDeviceContext: useCtx,
  DeviceProvider,
};

export {useCtx as useDeviceContext, DeviceProvider};
export default DeviceContext;
