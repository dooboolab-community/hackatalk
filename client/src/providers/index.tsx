export * from './AuthProvider';
export * from './DeviceProvider';
export * from './ProfileModalProvider';
