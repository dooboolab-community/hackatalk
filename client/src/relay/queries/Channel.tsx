import {graphql} from 'react-relay';

export const channelsQuery = graphql`
  query ChannelsQuery($first: Int!, $after: String, $withMessage: Boolean) {
    ...MainChannelComponent_channel
      @arguments(first: $first, after: $after, withMessage: $withMessage)
  }
`;

export const findOrCreatePrivateChannel = graphql`
  mutation ChannelFindOrCreatePrivateChannelMutation($peerUserIds: [String!]!) {
    findOrCreatePrivateChannel(peerUserIds: $peerUserIds) {
      id
      name
      channelType
    }
  }
`;
