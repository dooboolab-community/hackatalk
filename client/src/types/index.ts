import { StyleProp, TextStyle } from 'react-native';

import { SFC } from 'react';
import { User } from './graphql';

export enum MessageType {
  Message,
  Photo,
  File,
}

export interface Friend extends User {
  isFriend?: boolean;
  friendSince: Date;
}

interface IconProps {
  style?: StyleProp<TextStyle>;
  width?: number | string;
  height?: number | string;
  children?: never;
}

export type IconType = SFC<IconProps>;

export enum ThemeType {
  LIGHT = 'LIGHT',
  DARK = 'DARK',
}
