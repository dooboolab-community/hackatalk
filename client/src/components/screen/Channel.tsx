import * as Notifications from 'expo-notifications';

import type {
  ChannelsQuery,
  ChannelsQueryResponse,
  ChannelsQueryVariables,
} from '../../__generated__/ChannelsQuery.graphql';
import { FlatList, Platform, TouchableOpacity, View } from 'react-native';
import React, { FC, Suspense, useEffect, useMemo, useState } from 'react';
import {
  graphql,
  useLazyLoadQuery,
  usePaginationFragment,
  useQueryLoader,
} from 'react-relay/hooks';

import { AdMobBanner } from 'expo-ads-admob';
import { Channel } from '../../types/graphql';
import type {
  ChannelComponent_channel$key,
} from '../../__generated__/ChannelComponent_channel.graphql';
import { ChannelLastMessageQuery } from '../../__generated__/ChannelLastMessageQuery.graphql';
import ChannelListItem from '../shared/ChannelListItem';
import EmptyListItem from '../shared/EmptyListItem';
import { LoadingIndicator } from 'dooboo-ui';
import { MainStackNavigationProps } from '../navigation/MainStackNavigator';
import { SvgPlus } from '../../utils/Icons';
import { getString } from '../../../STRINGS';
import styled from 'styled-components/native';
import { useNavigation } from '@react-navigation/native';
import { useThemeContext } from '@dooboo-ui/theme';

const Container = styled.View`
  flex: 1;
  background: ${({ theme }): string => theme.background};
  flex-direction: column;
  align-items: center;
  justify-content: center;
`;

const Fab = styled.View`
  width: 56px;
  height: 56px;
  border-radius: 28px;
  justify-content: center;
  align-items: center;
  background: ${({ theme }): string => theme.fab};
`;

const ITEM_CNT = 20;

const channelsQuery = graphql`
  query ChannelsQuery(
    $first: Int
    $last: Int
    $before: String
    $after: String
    $withMessage: Boolean
  ) {
    ...ChannelComponent_channel @arguments(
      first: $first
      last: $last
      before: $before
      after: $after
      withMessage: $withMessage
    )
  }
`;

const channelsPaginationFragment = graphql`
  fragment ChannelComponent_channel on Query
    @argumentDefinitions(
      first: {type: "Int"}
      last: {type: "Int"}
      before: {type: "String"}
      after: {type: "String"}
      withMessage: {type: "Boolean"}
    )
    @refetchable(queryName: "Channels") {
      channels(
        first: $first
        last: $last
        before: $before
        after: $after
        withMessage: $withMessage
      )
      @connection(key: "ChannelComponent_channels" filters: ["withMessage"]) {
        edges {
          cursor
          node {
            id
            channelType
            name
            memberships(excludeMe: true) {
              user {
                name
                nickname
                thumbURL
                photoURL
              }
            }
            lastMessage {
              id
              messageType
              text
              imageUrls
              fileUrls
              createdAt
            }
          }
        }
        pageInfo {
          hasNextPage
          endCursor
        }
      }
    }
`;

const lastMessageQuery = graphql`
  query ChannelLastMessageQuery($messageId: String!) {
    message(id: $messageId) {
      id
      messageType
      text
      imageUrls
      fileUrls
      createdAt
      sender {
        id
      }
      channel {
        id
        lastMessage {
          id
        }
      }
    }
  }
`;

interface ChannelProps {
  channel: ChannelComponent_channel$key;
  searchArgs: ChannelsQueryVariables;
}

const ChannelsFragment: FC<ChannelProps> = ({
  channel,
  searchArgs,
}) => {
  const navigation = useNavigation();

  const {
    data,
    loadNext,
    isLoadingNext,
    refetch,
  } = usePaginationFragment<ChannelsQuery, ChannelComponent_channel$key>(
    channelsPaginationFragment,
    channel,
  );

  const [, loadLastMessage] = useQueryLoader<ChannelLastMessageQuery>(lastMessageQuery);
  const [bannerError, setBannerError] = useState<boolean>(false);

  useEffect(() => {
    const subscription = Notifications.addNotificationResponseReceivedListener(async (response) => {
      const messageId = JSON.parse(response.notification.request.content.data.data as string).messageId;

      if (typeof messageId === 'string') {
        loadLastMessage({ messageId });
      }
    });

    // Add notification handler.
    const responseListener = Notifications.addNotificationReceivedListener((event) => {
      const messageId = JSON.parse(event.request.content.data.data as string).messageId;

      loadNext(ITEM_CNT);

      if (typeof messageId === 'string') {
        loadLastMessage({ messageId });
      }
    });

    // Clean up : remove notification handler.
    return () => {
      Notifications.removeNotificationSubscription(responseListener);
      subscription.remove();
    };
  }, []);

  const onEndReached = (): void => {
    loadNext(ITEM_CNT);
  };

  const renderItem = ({
    item,
    index,
  }: {
    item: { node: Channel, cursor: string };
    index: number;
  }): React.ReactElement => {
    if (!item.node?.memberships || !item.node?.memberships.length) return <View key={index}/>;

    return (
      <ChannelListItem
        testID={`list-item-${index}`}
        key={index.toString()}
        item={item.node}
        onPress={(): void => {
          navigation.navigate('Message', {
            channel: item.node,
            users: item.node?.memberships?.map((membership) => membership?.user),
          });
        }}
      />
    );
  };

  const channels = useMemo(() => {
    return data?.channels?.edges ?? [];
  }, [data?.channels?.edges]);

  return <FlatList
    style={{
      alignSelf: 'stretch',
    }}
    contentContainerStyle={
      channels.length === 0
        ? {
          flex: 1,
          alignSelf: 'stretch',
          alignItems: 'center',
          justifyContent: 'center',
        }
        : null
    }
    keyExtractor={(_, index): string => index.toString()}
    // @ts-ignore
    data={channels}
    renderItem={renderItem}
    ListHeaderComponent={
      !bannerError
        ? Platform.select({
          android: <AdMobBanner
            bannerSize="fullBanner"
            // adUnitID="ca-app-pub-3940256099942544/6300978111"
            adUnitID="ca-app-pub-7837089095803162/8109702961"
            onDidFailToReceiveAdWithError={() => setBannerError(true)}
          />,
          ios: <AdMobBanner
            bannerSize="fullBanner"
            // adUnitID="ca-app-pub-3940256099942544/2934735716"
            adUnitID="ca-app-pub-7837089095803162/5084068464"
            onDidFailToReceiveAdWithError={() => setBannerError(true)}
          />,
        })
        : null
    }
    ListEmptyComponent={
      <EmptyListItem>{getString('NO_CHANNELLIST')}</EmptyListItem>
    }
    ListFooterComponent={
      <View style={{ height: 60 }} />
    }
    refreshing={isLoadingNext}
    onRefresh={() => {
      refetch(searchArgs, { fetchPolicy: 'network-only' });
    }}
    onEndReachedThreshold={0.1}
    onEndReached={onEndReached}
  />;
};

interface ContentProps {
  searchArgs: ChannelsQueryVariables;
}

const ContentContainer: FC<ContentProps> = ({
  searchArgs,
}) => {
  const data: ChannelsQueryResponse =
    useLazyLoadQuery<ChannelsQuery>(
      channelsQuery,
      searchArgs,
      { fetchPolicy: 'store-or-network' },
    );

  return <ChannelsFragment
    channel={data}
    searchArgs={searchArgs}
  />;
};

interface Props {
  navigation: MainStackNavigationProps<'Message'>;
}

const Screen: FC<Props> = () => {
  const { theme } = useThemeContext();
  const navigation = useNavigation();

  const searchArgs: ChannelsQueryVariables = {
    first: ITEM_CNT,
    withMessage: true,
  };

  return (
    <Container>
      <Suspense fallback={<LoadingIndicator/>}>
        <ContentContainer searchArgs={searchArgs}/>
      </Suspense>
      <TouchableOpacity
        activeOpacity={0.65}
        style={{
          position: 'absolute',
          right: 20,
          bottom: 44,
        }}
        onPress={(): void => navigation.navigate('ChannelCreate')}
      >
        <Fab>
          <SvgPlus fill={theme.background}/>
        </Fab>
      </TouchableOpacity>
    </Container>
  );
};

export default Screen;
