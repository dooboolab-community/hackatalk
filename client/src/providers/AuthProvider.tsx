import React, {useReducer} from 'react';

import {User} from '../types/graphql';
import createCtx from '../utils/createCtx';

export interface State {
  user: User | undefined;
}

interface Context {
  state: State;
  setUser(user: User | undefined): void;
}

const [useCtx, Provider] = createCtx<Context>();

// eslint-disable-next-line no-shadow
export enum ActionType {
  SetUser = 'set-user',
}

type SetUserAction = {
  type: ActionType.SetUser;
  payload: {user: User};
};
type Action = SetUserAction;

interface Props {
  children?: React.ReactElement;
  initialAuthUser?: User;
}

type Reducer = (state: State, action: Action) => State;

const setUser = (dispatch: React.Dispatch<SetUserAction>) => (
  authUser: User,
): void => {
  dispatch({
    type: ActionType.SetUser,
    payload: {user: authUser},
  });
};

const initialState: State = {
  user: undefined,
};

// eslint-disable-next-line default-param-last
const reducer: Reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.SetUser:
      return {...action.payload};
    default:
      return state;
  }
};

function AuthProvider(props: Props): React.ReactElement {
  const initialAuthState = {
    user: props.initialAuthUser,
  };

  const [state, dispatch] = useReducer<Reducer>(reducer, initialAuthState);

  const actions = {
    setUser: setUser(dispatch),
  };

  return <Provider value={{state, ...actions}}>{props.children}</Provider>;
}

const AuthContext = {
  useAuthContext: useCtx,
  AuthProvider,
};

export {useCtx as useAuthContext, AuthProvider};
export default AuthContext;
