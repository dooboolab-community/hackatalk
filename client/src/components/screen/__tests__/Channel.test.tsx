import 'react-native';

import * as React from 'react';

import { cleanup, fireEvent, render } from '@testing-library/react-native';
import { createTestElement, createTestProps } from '../../../../test/testUtils';

import Channel from '../Channel';
import { Channel as ChannelType } from '../../../types/graphql';
import { MockPayloadGenerator } from 'relay-test-utils';
import { environment } from '../../../providers';

const component = createTestElement(
  <Channel {...createTestProps()} />,
);

jest.mock('@react-navigation/native', () => {
  return {
    useNavigation: (): Record<string, unknown> => {
      return {
        navigate: jest.fn(),
      };
    },
  };
});

describe('[Channel] screen', () => {
  beforeEach(() => {
    environment.mockClear();
  });

  it('renders without crashing', () => {
    environment.mock.queueOperationResolver((operation) => {
      return MockPayloadGenerator.generate(operation, {
        Channel: (_, generateId): ChannelType => ({
          id: `user-${generateId()}`,
          name: 'HackaTalk',
          channelType: 'private',
          messages: {
            edges: [],
            pageInfo: {
              hasNextPage: false,
              hasPreviousPage: false,
            },
          },
        }),
      });
    });

    const json = render(component).toJSON();

    expect(json).toBeTruthy();
    expect(json).toMatchSnapshot();
  });

  describe('interactions', () => {
    beforeEach(() => {
      environment.mock.queueOperationResolver((operation) => {
        return MockPayloadGenerator.generate(operation, {
          Channel: (_, generateId): ChannelType => ({
            id: `user-${generateId()}`,
            name: 'HackaTalk',
            channelType: 'private',
            messages: {
              edges: [],
              pageInfo: {
                hasNextPage: false,
                hasPreviousPage: false,
              },
            },
          }),
        });
      });
    });

    it('should simulate onPress', () => {
      const { getByTestId } = render(component);
      const btn = getByTestId('list-item-0');

      fireEvent.press(btn);
    });
  });
});
