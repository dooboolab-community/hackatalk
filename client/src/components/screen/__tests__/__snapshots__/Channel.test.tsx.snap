// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Channel] screen renders without crashing 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#ffffff",
        "flexBasis": 0,
        "flexDirection": "column",
        "flexGrow": 1,
        "flexShrink": 1,
        "justifyContent": "center",
      },
    ]
  }
>
  <RCTScrollView
    ListEmptyComponent={
      <Shared
        style={
          Object {
            "alignItems": "center",
            "alignSelf": "stretch",
            "flex": 1,
            "justifyContent": "center",
            "width": "100%",
          }
        }
      >
        Start a Chat
      </Shared>
    }
    ListFooterComponent={
      <View
        style={
          Object {
            "height": 60,
          }
        }
      />
    }
    contentContainerStyle={null}
    data={
      Array [
        Object {
          "cursor": "<mock-value-for-field-\\"cursor\\">",
          "node": Object {
            "__typename": "Channel",
            "channelType": "private",
            "id": "user-1",
            "lastMessage": Object {
              "createdAt": "<mock-value-for-field-\\"createdAt\\">",
              "fileUrls": "<mock-value-for-field-\\"fileUrls\\">",
              "id": "<Message-mock-id-2>",
              "imageUrls": "<mock-value-for-field-\\"imageUrls\\">",
              "messageType": "<mock-value-for-field-\\"messageType\\">",
              "text": "<mock-value-for-field-\\"text\\">",
            },
            "memberships": Array [
              Object {
                "user": Object {
                  "name": "<mock-value-for-field-\\"name\\">",
                  "nickname": "<mock-value-for-field-\\"nickname\\">",
                  "photoURL": "<mock-value-for-field-\\"photoURL\\">",
                  "thumbURL": "<mock-value-for-field-\\"thumbURL\\">",
                },
              },
            ],
            "name": "HackaTalk",
          },
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={0.1}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onRefresh={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshControl={
      <RefreshControlMock
        onRefresh={[Function]}
        refreshing={false}
      />
    }
    refreshing={false}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "alignSelf": "stretch",
      }
    }
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <RCTRefreshControl />
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "width": "100%",
            }
          }
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="list-item-0"
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "#ffffff",
                    "borderBottomWidth": 1,
                    "borderColor": "rgb(229, 229, 229)",
                    "flexDirection": "row",
                    "height": 92,
                    "justifyContent": "space-between",
                  },
                ]
              }
            >
              <View
                style={
                  Object {
                    "marginHorizontal": 20,
                  }
                }
              >
                <Image
                  source={
                    Object {
                      "uri": "<mock-value-for-field-\\"thumbURL\\">",
                    }
                  }
                  style={
                    Array [
                      Object {
                        "borderRadius": 20,
                        "height": 40,
                        "width": 40,
                      },
                    ]
                  }
                />
                <View />
              </View>
              <View
                style={
                  Array [
                    Object {
                      "flexBasis": 0,
                      "flexDirection": "column",
                      "flexGrow": 1,
                      "flexShrink": 1,
                      "paddingRight": 20,
                    },
                  ]
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "alignItems": "center",
                        "flexDirection": "row",
                        "justifyContent": "space-between",
                      },
                    ]
                  }
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "color": "#151A25",
                          "fontSize": 14,
                          "fontWeight": "bold",
                        },
                      ]
                    }
                  >
                    &lt;mock-value-for-field-"nickname"&gt;
                  </Text>
                </View>
                <View
                  style={
                    Array [
                      Object {
                        "alignItems": "center",
                        "flexDirection": "row",
                        "justifyContent": "space-between",
                        "marginTop": 8,
                      },
                    ]
                  }
                >
                  <Text
                    lastMessageCnt={0}
                    numberOfLines={2}
                    style={
                      Array [
                        Object {
                          "color": "#151A25",
                          "fontSize": 12,
                          "maxWidth": 150,
                        },
                      ]
                    }
                  >
                    &lt;mock-value-for-field-"text"&gt;
                  </Text>
                  <Text
                    style={
                      Array [
                        Object {
                          "color": "#1E6EFA",
                          "fontSize": 12,
                          "textAlign": "right",
                        },
                      ]
                    }
                  >
                    Invalid date
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
      >
        <View
          style={
            Object {
              "height": 60,
            }
          }
        />
      </View>
    </View>
  </RCTScrollView>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "bottom": 44,
        "opacity": 1,
        "position": "absolute",
        "right": 20,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#1E6EFA",
            "borderRadius": 28,
            "height": 56,
            "justifyContent": "center",
            "width": 56,
          },
        ]
      }
    >
      <SvgMock
        fill="#ffffff"
      />
    </View>
  </View>
</View>
`;
